sdks/python/apache_beam/ml/anomaly/transforms.py

```diff
@@ -0,0 +1,381 @@
+
+    Returns:
+      float: The anomaly score predicted by the model.
+    """
+    assert self._underlying
+    if self._underlying._features is not None:
+      x = beam.Row(**{f: getattr(data, f) for f in self._underlying._features})
+    else:
+      x = beam.Row(**data._asdict())
+
+    # score the incoming data using the existing model
+    y_pred = self._underlying.score_one(x)
+
+    # then update the model with the same data
+    self._underlying.learn_one(x)
+
+    return y_pred
+
+  def process(
+      self,
+      element: KeyedInputT,
+      model_state=beam.DoFn.StateParam(MODEL_STATE_INDEX),
+      **kwargs) -> Iterable[KeyedOutputT]:
+
+    model_state = typing.cast(ReadModifyWriteRuntimeState, model_state)
+    k1, (k2, data) = element
+    self._underlying: AnomalyDetector = model_state.read()
+    if self._underlying is None:
+      self._underlying = typing.cast(
+          AnomalyDetector, Specifiable.from_spec(self._detector_spec))
+
+    yield k1, (k2,
+               AnomalyResult(
+                   example=data,
+                   predictions=[AnomalyPrediction(
+                       model_id=self._underlying._model_id,
+                       score=self.score_and_learn(data))]))
+
+    model_state.write(self._underlying)
+
+
+class RunScoreAndLearn(beam.PTransform[beam.PCollection[KeyedInputT],
+                                       beam.PCollection[KeyedOutputT]]):
+  """Applies the _ScoreAndLearnDoFn to a PCollection of data.
+
+  This PTransform scores and learns from data points using an anomaly
+  detection model.
+
+  Args:
+    detector: The anomaly detection model to use.
+  """
+  def __init__(self, detector: AnomalyDetector):
+    self._detector = detector
+
+  def expand(
+      self,
+      input: beam.PCollection[KeyedInputT]) -> beam.PCollection[KeyedOutputT]:
+    return input | beam.ParDo(_ScoreAndLearnDoFn(self._detector.to_spec()))
+
+
+class RunThresholdCriterion(beam.PTransform[beam.PCollection[KeyedOutputT],
+                                            beam.PCollection[KeyedOutputT]]):
+  """Applies a threshold criterion to anomaly detection results.
+
+  This PTransform applies a `ThresholdFn` to the anomaly scores in
+  `AnomalyResult` objects, updating the prediction labels. It handles both
+  stateful and stateless `ThresholdFn` implementations.
+
+  Args:
+    threshold_criterion: The `ThresholdFn` to apply.
+  """
+  def __init__(self, threshold_criterion):
+    self._threshold_fn = threshold_criterion
+
+  def expand(
+      self,
+      input: beam.PCollection[KeyedOutputT]) -> beam.PCollection[KeyedOutputT]:
+    if self._threshold_fn:
+      if self._threshold_fn.is_stateful:
+        ret = (
+            input
+            | beam.ParDo(StatefulThresholdDoFn(self._threshold_fn.to_spec())))
+      else:
+        ret = (
+            input
+            | beam.ParDo(StatelessThresholdDoFn(self._threshold_fn.to_spec())))
+    else:
+      ret = input
+
+    return ret
```

Comment:
This can be simplified with early returns
```suggestion
    if not self._threshold_fn:
      return input
    if self._threshold_fn.is_stateful:
      return (
          input
          | beam.ParDo(StatefulThresholdDoFn(self._threshold_fn.to_spec())))
    return (
        input
        | beam.ParDo(StatelessThresholdDoFn(self._threshold_fn.to_spec())))
```

Reply:
But also, its a bit odd that we're even accepting a threshold_fn of None - seems like maybe we should exclude that with typing and just not expand this in that case

Reply:
Fixed.